### PR TITLE
Fold System.IO.FileSystem into CoreLib

### PR DESCRIPTION
**Pull request by adamsitnik**
Contributes to #2138

### CODE REVIEW

**Comment by adamsitnik** [comment]
@jkotas `System.Linq` dependency was not a problem, as only `.ToArray()` was being used and I was able to reference `EnumerableHelpers` to get it working. It turned out that we also have a dependency to `Queue<T>` and `Stack<T>` which I had to move as well. Is that acceptable?

I followed the pattern that you have used for `System.IO`, but I got stuck with the following error:

```log
/home/adam/projects/runtime/src/libraries/Microsoft.NETCore.Platforms/src/runtimeGroups.props(277,5): error MSB4018: The "GenerateRuntimeGraph" task failed unexpectedly. [/home/adam/projects/runtime/src/libraries/Microsoft.NETCore.Platforms/src/Microsoft.NETCore.Platforms.csproj]
/home/adam/projects/runtime/src/libraries/Microsoft.NETCore.Platforms/src/runtimeGroups.props(277,5): error MSB4018: System.TypeLoadException: Could not load type 'System.IO.File' from assembly 'System.Runtime, Version=6.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'. [/home/adam/projects/runtime/src/libraries/Microsoft.NETCore.Platforms/src/Microsoft.NETCore.Platforms.csproj]
/home/adam/projects/runtime/src/libraries/Microsoft.NETCore.Platforms/src/runtimeGroups.props(277,5): error MSB4018:    at Microsoft.NETCore.Platforms.BuildTasks.GenerateRuntimeGraph.Execute() [/home/adam/projects/runtime/src/libraries/Microsoft.NETCore.Platforms/src/Microsoft.NETCore.Platforms.csproj]
/home/adam/projects/runtime/src/libraries/Microsoft.NETCore.Platforms/src/runtimeGroups.props(277,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute() [/home/adam/projects/runtime/src/libraries/Microsoft.NETCore.Platforms/src/Microsoft.NETCore.Platforms.csproj]
/home/adam/projects/runtime/src/libraries/Microsoft.NETCore.Platforms/src/runtimeGroups.props(277,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskBuilder.ExecuteInstantiatedTask(ITaskExecutionHost taskExecutionHost, TaskLoggingContext taskLoggingContext, TaskHost taskHost, ItemBucket bucket, TaskExecutionMode howToExecuteTask) [/home/adam/projects/runtime/src/libraries/Microsoft.NETCore.Platforms/src/Microsoft.NETCore.Platforms.csproj]
```

@jkotas is there any chance that you could take a look at my changes and see what I am doing wrong?

**Comment by jkotas** [comment]
> `Stack<T>`

The difference between Stack<T> and List<T> is miniscule. Both are growable arrays, with just slightly different methods exposed. The use of `Stack<T>` can be trivially replaced by `List<T>` if you would like to avoid introducing the dependency on Stack.

I think it is ok to move the Queue over.

**Comment by jkotas** [comment]
>  I got stuck with the following error:

I do not know what can be causing this error. It looks like a bug in the GenerateRuntimeGraph msbuild task to me. It shoud be using the live-built assets, but it is using cached assets instead. You may want to do a clean build to make sure that it is a real problem.

**Comment by adamsitnik** [comment]
@ViktorHofer @Anipik In this PR I am trying to move `System.IO.File` from `System.IO.FileSystem` to `System.PrivateCoreLib`. I got stuck with the following issue:

```log
C:\Projects\runtime\src\libraries\Microsoft.NETCore.Platforms\src\runtimeGroups.props(277,5): error MSB4018: The "GenerateRuntimeGraph" task failed unexpectedly. [C:\Projects\runtime\src\libraries\Microsoft.NETCore.Platforms\src\Microsoft.NETCore.Platforms.csproj]
C:\Projects\runtime\src\libraries\Microsoft.NETCore.Platforms\src\runtimeGroups.props(277,5): error MSB4018: System.TypeLoadException: Could not load type 'System.IO.File' from assembly 'System.Runtime, Version=6.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'. [C:\Projects\runtime\src\libraries\Microsoft.NETCore.Platforms\src\Microsoft.NETCore.Platforms.csproj]
C:\Projects\runtime\src\libraries\Microsoft.NETCore.Platforms\src\runtimeGroups.props(277,5): error MSB4018:    at Microsoft.NETCore.Platforms.BuildTasks.GenerateRuntimeGraph.Execute() [C:\Projects\runtime\src\libraries\Microsoft.NETCore.Platforms\src\Microsoft.NETCore.Platforms.csproj]
C:\Projects\runtime\src\libraries\Microsoft.NETCore.Platforms\src\runtimeGroups.props(277,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute() [C:\Projects\runtime\src\libraries\Microsoft.NETCore.Platforms\src\Microsoft.NETCore.Platforms.csproj]
C:\Projects\runtime\src\libraries\Microsoft.NETCore.Platforms\src\runtimeGroups.props(277,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskBuilder.ExecuteInstantiatedTask(ITaskExecutionHost taskExecutionHost, TaskLoggingContext taskLoggingContext, TaskHost taskHost, ItemBucket bucket, TaskExecutionMode howToExecuteTask) [C:\Projects\runtime\src\libraries\Microsoft.NETCore.Platforms\src\Microsoft.NETCore.Platforms.csproj]
```

I've verified @jkotas suggestion using a clean build:

>  It looks like a bug in the GenerateRuntimeGraph msbuild task to me. It shoud be using the live-built assets, but it is using cached assets instead.

 and it's true that the msbuild task is using the wrong assets. In my case it's using `C:\Projects\runtime\.dotnet\shared\Microsoft.NETCore.App\6.0.0-preview.3.21201.4\System.Private.CoreLib.dll` but with the information that `System.IO.File` has been already moved to `System.Private.CoreLib`

@ViktorHofer @Anipik do you have any suggestions about how this could be solved (or worked around)?

File: src/libraries/Common/src/System/Collections/Generic/EnumerableHelpers.cs [hunk 1]
```diff
@@ -8,6 +8,13 @@ namespace System.Collections.Generic
     /// </summary>
     internal static partial class EnumerableHelpers
```
Comment (jkotas) [inline src/libraries/Common/src/System/Collections/Generic/EnumerableHelpers.cs:9]: Is this still needed after the recent commit?

Comment (adamsitnik) [reply src/libraries/Common/src/System/Collections/Generic/EnumerableHelpers.cs:9]: Yes, as `Queue` uses it:

https://github.com/dotnet/runtime/blob/3bd75ba9f1a6a7a8727e53a1a15589ff3a974f30/src/libraries/System.Collections/src/System/Collections/Generic/Queue.cs#L57

Comment (jkotas) [reply src/libraries/Common/src/System/Collections/Generic/EnumerableHelpers.cs:9]: That should not be needed for `MS_IO_REDIST`. Can the ifdefs in this file be dropped?

**Comment by ViktorHofer** [comment]
> and it's true that the msbuild task is using the wrong assets. In my case it's using C:\Projects\runtime\.dotnet\shared\Microsoft.NETCore.App\6.0.0-preview.3.21201.4\System.Private.CoreLib.dll but with the information that System.IO.File has been already moved to System.Private.CoreLib

@ericstj was the one who brought the task over from arcade into dotnet/runtime so he's the right person to provide an answer. But to give you some more data meanwhile, the failing task isn't expected to run against the live shared framework as it contributes to the build itself. This is similar to msbuild and other tools that we use to build the repository which also run on a last-known-good shipped version of the shared framework (currently 6.0 Preview 3).

**Comment by jkotas** [comment]
> This is similar to msbuild and other tools that we use to build the repository which also run on a last-known-good shipped version of the shared framework

That means these tools should build against the last-known-good shipped references, and not against the live references.

**Comment by ViktorHofer** [comment]
> That means these tools should build against the last-known-good shipped references, and not against the live references.

Exactly. That needs to be fixed as the project currently targets the live references.

**Comment by ViktorHofer** [comment]
Had some minutes. Should be fixed with https://github.com/dotnet/runtime/pull/53290. cc @ericstj

**Comment by ViktorHofer** [comment]
Curious, what's the benefit of moving System.IO.FileSystem down into CoreLib? And more generally asking what are your thoughts about which types belong into CoreLib and which don't? With time, will more and more types move down into CoreLib? I think that has already been happening over the last years?

**Comment by jkotas** [comment]
> Curious, what's the benefit of moving System.IO.FileSystem down into CoreLib?

Discussed at https://github.com/dotnet/runtime/pull/53168#issuecomment-847242457

> And more generally asking what are your thoughts about which types belong into CoreLib and which don't?

I think the generalized rule is - when other types in CoreLib need to depend on the type, the type has to be in CoreLib.

**Comment by ViktorHofer** [comment]
> I think the generalized rule is - when other types in CoreLib need to depend on the type, the type has to be in CoreLib.

Doesn't that also mean that we will move more and more types down into CoreLib and it becomes an ever growing assembly? Should all low level APIs like FileSystem belong into CoreLib?

File: src/libraries/System.IO.IsolatedStorage/ref/System.IO.IsolatedStorage.csproj [hunk 2]
```diff
@@ -7,7 +7,6 @@
     <Compile Include="System.IO.IsolatedStorage.cs" />
   </ItemGroup>
   <ItemGroup>
-    <ProjectReference Include="..\..\System.IO.FileSystem\ref\System.IO.FileSystem.csproj" />
     <ProjectReference Include="..\..\System.IO.FileSystem.Primitives\ref\System.IO.FileSystem.Primitives.csproj" />
```
Comment (jkotas) [inline src/libraries/System.IO.IsolatedStorage/ref/System.IO.IsolatedStorage.csproj:10]: You can delete `System.IO.FileSystem.Primitives.csproj` refs too while you are on it. System.IO.FileSystem.Primitives does not have anything real it in.

File: src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Unix.cs [hunk 3]
```diff
@@ -302,7 +302,8 @@ public static void CreateDirectory(string fullPath)
             string errorString = fullPath;
             while (stackDir.Count > 0)
```
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Unix.cs:303]:
```
for (int i = stackDir.Count - 1; i >= 0; i--)
{
    string name = stackDir[i];
...
```
I do not think you need to spend time on actually removing the items from the list.

Comment (jkotas) [reply src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Unix.cs:303]: Or you can get rid of the list alltogether, and just go through the path and look for directory seperators.

File: src/libraries/System.Private.CoreLib/src/System/IO/File.cs [hunk 4]
```diff
@@ -352,7 +352,11 @@ public static byte[] ReadAllBytes(string path)
                 {
                     int n = fs.Read(bytes, index, count);
                     if (n == 0)
+#if !MS_IO_REDIST
+                        ThrowHelper.ThrowEndOfFileException();
```
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/IO/File.cs:356]: Can we add ThrowHelper to `MS_IO_REDIST` instead of adding ifdef?

**Review by jkotas — APPROVED** [review]
LGTM otherwise. Thanks!

**Comment by ViktorHofer** [comment]
Given that System.IO.FileSystem only contains type forwards now, should we also remove the references to it in projects that target `$(NetCoreAppCurrent)` as i.e. `System.IO.File` is now also exposed in `System.Runtime`?

**Comment by ViktorHofer** [comment]
cc @ericstj

**Comment by jkotas** [comment]
> Given that System.IO.FileSystem only contains type forwards now, should we also remove the references to it in projects that target $(NetCoreAppCurrent) as i.e. System.IO.File is now also exposed in System.Runtime?

We cannot. It would break references to existing packages that still reference System.IO.FileSystem. (Note that we have quite refererence .dlls in netcoreapp that only contain forwarders.)

>  I'm seeing issues when referencing both assemblies at the same time as the types are overlapping.

If you are seeing conflicting types, I think you are using inconsistent set of references - some are from live netcoreapp6 and some are from older netcoreapp.

**Comment by ViktorHofer** [comment]
> If you are seeing conflicting types, I think you are using inconsistent set of references - some are from live netcoreapp6 and some are from older netcoreapp.

This was because of a non clean build 🤦 - sorry

> We cannot. It would break references to existing packages that still reference System.IO.FileSystem. (Note that we have quite refererence .dlls in netcoreapp that only contain forwarders.)

**Pull request merged**